### PR TITLE
Host Metrics process scraper

### PR DESCRIPTION
**Pull request by james-bebbington**
**Link to tracking Issue:**
https://github.com/open-telemetry/opentelemetry-collector/issues/847

**Description:**
Added process scraper to the hostmetricsreceiver which currently scrapes cpu, memory & disk usage stats using gopsutil. Also includes the ability to include/exclude processes by name using the existing `filterset` config.

**Example Metrics:**
image
image
image
image

^ in most cases "command_line" is not empty, but its missing for a few system processes on Windows

**Benchmarks:**
image

### CODE REVIEW

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	processAccessors := make([]processHandle, len(processes))
+	for i, proc := range processes {
+		processAccessors[i] = &gopsProcessHandle{proc}
+	}
+	return processAccessors, nil
+}
```
Comment (james-bebbington) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:54]: This function exists only to convert the slice returned by `gopsutil/process.Processes()` into a slice of the interface type declared above so that we can write tests against this. I'm not aware of a better way to do this in Go but happy to learn if there is a better alternative.

File: go.mod [hunk 2]
```diff
@@ -42,7 +42,7 @@ require (
 	github.com/prometheus/prometheus v1.8.2-0.20190924101040-52e0504f83ea
 	github.com/rs/cors v1.6.0
 	github.com/securego/gosec v0.0.0-20200316084457-7da9f46445fd
-	github.com/shirou/gopsutil v2.20.4+incompatible
+	github.com/shirou/gopsutil v0.0.0-20200517204708-c89193f22d93 // c89193f22d9359848988f32aee972122bb2abdc2
```
Comment (james-bebbington) [inline go.mod:45]: `gopsutil` needed to be updated to include this relatively recent PR that improves the performance of reading process information on Windows: https://github.com/shirou/gopsutil/pull/862

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+type gopsProcessHandle struct {
+	*process.Process
+}
+
+func (p *gopsProcessHandle) GetPid() int32 {
+	return p.Pid
+}
+
+var getProcessHandles = getProcessHandlesInternal
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:41]: This var is a global state, which makes modifications unsafe e.g. if executing tests in parallel.
I'd recommend making it a field in the `scraper` struct, which is set to `getProcessHandlesInternal` by default, but can be overriden in tests.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+	waitStateLabelValue   = "wait"
+)
+
+var metricCPUUsageDescriptor = createMetricCPUUsageDescriptor()
+
+func createMetricCPUUsageDescriptor() pdata.MetricDescriptor {
+	descriptor := pdata.NewMetricDescriptor()
+	descriptor.InitEmpty()
+	descriptor.SetName("process/cpu/usage")
+	descriptor.SetDescription("Total CPU seconds broken down by different states")
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go:48]: nit: missing `.` at the end (for consistency with descriptors below)?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+	memoryInfo *process.MemoryInfoStat
+}
+
+type processIoCounters struct {
+	*processMetadata
+	ioCounters *process.IOCountersStat
+}
+
+// newProcessScraper creates a Process Scraper
+func newProcessScraper(_ context.Context, cfg *Config) (*scraper, error) {
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:66]: Do you actually need a context parameter if you never use it?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+// ScrapeMetrics
+func (s *scraper) ScrapeMetrics(ctx context.Context) (pdata.MetricSlice, error) {
+	_, span := trace.StartSpan(ctx, "processscraper.ScrapeMetrics")
+	defer span.End()
+
+	var errors []error
+	metrics := pdata.NewMetricSlice()
+
+	processes, err := s.getProcesses()
+	if len(processes) == 0 {
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:113]: It's more idiomatic to check for `if err != nil` first, and  result should be empty when err is not nil.
```go
if err != nil {
  return nil, err
} else if len(processes) == 0 {
  return metrics, nil
}
```

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+		return metrics, err
+	}
+
+	if err != nil {
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:117]: This condition should be impossible given the check for empty processes above, since result must be empty when err is not nil.
Is there a possibility of getting non-empty list of processes AND error at the same time?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+	cpuTimes := make([]*processTimes, 0, len(processes))
+	memoryInfo := make([]*processMemoryInfo, 0, len(processes))
+	ioCounters := make([]*processIoCounters, 0, len(processes))
+
+	for _, metadata := range processes {
+		times, err := metadata.handle.Times()
+		if err != nil {
+			errors = append(errors, fmt.Errorf("error reading process cpu times for process %q (pid %v): %v", metadata.name, metadata.pid, err))
+		} else {
+			cpuTimes = append(cpuTimes, &processTimes{metadata, times})
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:130]: nit: While shorter, setting fields in struct based on order is error prone. Probably okay here though given different types of fields.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+	}
+
+	cpuTimes := make([]*processTimes, 0, len(processes))
+	memoryInfo := make([]*processMemoryInfo, 0, len(processes))
+	ioCounters := make([]*processIoCounters, 0, len(processes))
+
+	for _, metadata := range processes {
+		times, err := metadata.handle.Times()
+		if err != nil {
+			errors = append(errors, fmt.Errorf("error reading process cpu times for process %q (pid %v): %v", metadata.name, metadata.pid, err))
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:128]: Shouldn't we append the empty / default item to `cpuTimes` here to preserve order of the processes?
AFAICT the index of items in each slice must exactly match the index of process in `processes`.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+		} else {
+			cpuTimes = append(cpuTimes, &processTimes{metadata, times})
+		}
+
+		mem, err := metadata.handle.MemoryInfo()
+		if err != nil {
+			errors = append(errors, fmt.Errorf("error reading process memory info for process %q (pid %v): %v", metadata.name, metadata.pid, err))
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:135]: What is intended behavior if `Time()` returned error above? Do we want to still record memory and IO counters?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 5 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: Removed

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 6 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: I've followed that approach where possible, but it's not so straightforward in a case like this where partial failure is possible (failed to get some processes, succeeded in getting others), and we want to propagate any partial errors up to be reported on in aggregate.

For now I've added a comment to the function that returns partial results to document this behaviour, but I'm happy to explore other options if this is strongly unidiomatic. I guess another option would be to store partial errors in a different type (i.e. not error)?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 9 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: The ordering of data in these slices doesn't matter, but the code was kinda hard to follow since we were obtaining the data for three different metrics and then initializaing them later. I've refactored this so each metric is scraped & initialized one-by-one, and its hopefully much easier to follow now.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 10 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: Yes that's right. Should be much clearer now as per the refactoring mentioned above.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 6 above)
Comment (nilebox) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: Probably returning a "result" struct rather than tuple would help to make it more idiomatic.
i.e. in that case error is part of the function result rather than an "exception"

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 7 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:118]: See the comment / discussion above which explains the reasoning for this. Note the additional check that was here before for processes being empty wasn't necessary. I've changed this now to simply read:

```go
processes, err := s.getProcesses()
if err != nil {
	errors = append(errors, err)
}
```

... which is closer to being idiomatic

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 3 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go]: While I do get your point, I feel like storing this in the scraper struct is a little bit less readable than storing it right above the "real" function here, and I'm not totally sure its worth the benefit of making modifications safe, given that it would only ever be overridden in tests (which don't run in parallel by default anwyay)

But I've made this change anyway 😄

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 6 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: Alright - I might leave that for a separate PR though as this style is used throughout the existing scrapers at the moment (in particular, in the result of the `ScrapeMetrics` function)

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 1 above)
Comment (bogdandrutu) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:52]: You can make the getProcessHandlesInternal a function pointer that you replace that pointer so you can return the initial struct not interface but you can mock the results. Does that work?

Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:52]: I don't think that will work because Go doesn't support covariance of slices.

```
cannot use processes (variable of type []*process.Process) as []processHandle value in return statement
```

Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:61]: I've added a second commit that refactors this function to return an interface that represents a slice (with `At()` & `Len()` funcs) so that we only need to wrap the returned slice once instead of wrapping each element. This seems to be a better trade off.

The `gopsutil/process.Process` struct has a Pid property rather than an accessor function, so I had to make accessing the Pid a function on the root struct which is slightly odd but doesn't effect readability too much.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/config.go [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+	// Exclude specifies a filter on the process names that should be excluded from the generated metrics.
+	// If neither `include` or `exclude` are set, process metrics will be generated for all processes.
+	Include MatchConfig `mapstructure:"include"`
+	Exclude MatchConfig `mapstructure:"exclude"`
+}
+
+type MatchConfig struct {
+	filterset.Config `mapstructure:",squash"`
+
+	Names []string `mapstructure:"names"`
```
Comment (bogdandrutu) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/config.go:36]: What is the Names slice here? why not using the "strict" match type for the config?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+
+const (
+	userStateLabelValue   = "user"
+	systemStateLabelValue = "system"
+	waitStateLabelValue   = "wait"
+)
+
+var metricCPUUsageDescriptor = createMetricCPUUsageDescriptor()
+
+func createMetricCPUUsageDescriptor() pdata.MetricDescriptor {
```
Comment (bogdandrutu) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go:44]: These are not constants. Maybe just utils.go? Not sure what is the right name.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 7 above)
Comment (bogdandrutu) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:120]:
```go
if err := scrapeAndAppendCPUUsageMetric(metrics, s.startTime, processes); err != nil {
  errors = append(errors, err)
}
```

Same everywhere where you have just an error. It is more compact and simpler to read (personal opinion).

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/config.go (hunk 11 above)
Comment (bogdandrutu) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/config.go:36]: I saw the current filters do not have the actual strings in the config. Will take care of that later.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go (hunk 12 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go:44]: Yea good point - maybe `metadata.go`? _matching the proposal https://github.com/open-telemetry/opentelemetry-collector/issues/985#issuecomment-632307672_

I'll leave this refactoring to another PR though as its the same across all the scrapers.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 7 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:120]: Updated

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go (hunk 12 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_constants.go]: See https://github.com/open-telemetry/opentelemetry-collector/pull/1074

File: receiver/hostmetricsreceiver/example_config.yaml [hunk 13]
```diff
@@ -12,6 +12,7 @@ receivers:
       disk:
       filesystem:
       network:
+      virtualmemory:
```
Comment (bogdandrutu) [inline receiver/hostmetricsreceiver/example_config.yaml:15]: This is process... Why virtualmem? Forgot to add process?

File: receiver/hostmetricsreceiver/internal/testutils.go [hunk 14]
```diff
@@ -45,3 +46,40 @@ func AssertInt64MetricLabelDoesNotExist(t *testing.T, metric pdata.Metric, index
...
+		value, ok := idps.At(i).LabelsMap().Get(labelName)
+		if ok && value.Value() != "" {
+			return
+		}
+	}
+
+	assert.Failf(t, "Missing label %q for all datapoints in metric %q", labelName, metric.MetricDescriptor().Name())
+}
+
+func AssertInt64MetricLabelHasValuesAllDataPoints(t *testing.T, metric pdata.Metric, labelName string, expectedValues []string) {
```
Comment (bogdandrutu) [inline receiver/hostmetricsreceiver/internal/testutils.go:62]: Consider a comment for this func. Maybe same for previous.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go (hunk 1 above)
Comment (bogdandrutu) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process.go:61]: Please comment the struct that it exists just for testing.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 15]
```diff
@@ -0,0 +1,352 @@
...
+
+	return metadata, nil
+}
+
+func getProcessName(proc processHandle) (string, error) {
+	if runtime.GOOS != "windows" {
+		return proc.Name()
+	}
+
+	// calling proc.Name() is currently prohibitively expensive on Windows, so use the exe name instead (mirrors psutil)
```
Comment (jrcamp) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:186]: Is it because it's a WMI call or other reasons?

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 16]
```diff
@@ -0,0 +1,352 @@
...
+}
+
+type processMemoryInfo struct {
+	*processMetadata
+	memoryInfo *process.MemoryInfoStat
+}
+
+type processIoCounters struct {
+	*processMetadata
+	ioCounters *process.IOCountersStat
```
Comment (jrcamp) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:64]: These structs are relatively small, would maybe embed processMetadata as a value instead and see if it improves benchmarks any.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 17]
```diff
@@ -0,0 +1,352 @@
...
+	}
+
+	return metrics, nil
+}
+
+// getProcesses returns a slice of processMetadata, including handles,
+// for all currently running processes. If errors occur obtaining
+// information for some processes, an error will be returned, but any
+// processes that were successfully obtained will still be returned.
+func (s *scraper) getProcesses() ([]*processMetadata, error) {
```
Comment (jrcamp) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:142]: Would try making doing processMetadata by value and seeing if it improves benchmark.

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 15 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:186]: No, in this case it calls https://docs.microsoft.com/en-us/windows/win32/api/tlhelp32/nf-tlhelp32-createtoolhelp32snapshot (using the TH32CS_SNAPPROCESS flag). I don't know why this is so slow but from my testing it was taking around 1s (on a machine running lots of processes)

File: receiver/hostmetricsreceiver/example_config.yaml (hunk 13 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/example_config.yaml:15]: I forgot to add `virtualmemory` before (thus adding it now 🤦‍♂️),

I intentionally did not add `process` to the default/example config, as process metrics are more expensive to compute and may often not be needed

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_metadata.go [hunk 18]
```diff
@@ -0,0 +1,82 @@
...
+
+var metricCPUUsageDescriptor = createMetricCPUUsageDescriptor()
+
+func createMetricCPUUsageDescriptor() pdata.MetricDescriptor {
+	descriptor := pdata.NewMetricDescriptor()
+	descriptor.InitEmpty()
+	descriptor.SetName("process/cpu/usage")
+	descriptor.SetDescription("Total CPU seconds broken down by different states.")
+	descriptor.SetUnit("s")
+	descriptor.SetType(pdata.MetricTypeCounterInt64)
```
[the remark this answers is not in the document]
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_metadata.go]: Yes, good point - this was losing precision 🤦

I will need to change this for "host/cpu" as well

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 16 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go]: I've committed this change, though it's too hard to tell if made any impact to the benchmarks (at least on Windows) as this change is negligible compared to work done in the syscalls (which are not very consistent).

**Comment by james-bebbington** [comment]
> How many processes are running on the machine you're benchmarking with? Any idea what the max we may see in production? There are optimizations we could do but I think we need some realistic workloads running to validate against.

The benchmark above was from running on my local machine with ~300 processes. I'll get back to you with some estimates on what workloads we expect

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go (hunk 15 above)
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:197]: Could we maybe call this instead? https://docs.microsoft.com/en-us/windows/win32/api/psapi/nf-psapi-getmodulebasenamew

Comment (jrcamp) [reply receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:197]: Ah, yeah CreateToolhelp32Snapshot sounds rather expensive. What you have seems fine to me.

**Comment by jrcamp** [comment]
@james-bebbington is this ready for merge now?

**Comment by james-bebbington** [comment]
I forgot to leave a comment here to say this is waiting on a change to treat each "process" as a resource instead of as labels on these metrics.

I will update this PR once https://github.com/open-telemetry/opentelemetry-specification/pull/635 is merged.

**Comment by james-bebbington** [comment]
I've now updated this to create a separate resource (and associated metrics) for each process. This did end up requiring some fairly significant changes: the existing Scraper API only supported returning a MetricSlice, so I created a separate API to use for the Process Scraper that returns a ResourceMetricsSlice.

Apologies for this PR getting a bit large - I've separated the refactoring around how resources are handled to a separate commit, so for those who've already taken a look, you can just review the final commit.

FYI @jrcamp @nilebox  @bogdandrutu

File: receiver/hostmetricsreceiver/internal/scraper.go [hunk 19]
```diff
@@ -39,15 +50,32 @@ type Scraper interface {
...
-		cfg Config) (Scraper, error)
+	CreateMetricsScraper(ctx context.Context, logger *zap.Logger, cfg Config) (Scraper, error)
+}
+
+// ResourceScraper gathers metrics from a low-level resource such as
+// a process.
+type ResourceScraper interface {
+	BaseScraper
+
+	// ScrapeMetrics returns relevant scraped metrics per resource.
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper.go:65]: No suggestion, just trying to understand the problem better, since this seems to be the key change here.

AFAICT the main difference between `Scraper` and `ResourceScraper` is that Scraper returns metrics for a single fixed resource (host), while `ResourceScraper` dynamically gathers a list of multiple processes, generates a resource per process and returns metrics for each of them.

So in that sense, `Scraper` is just a special case of `ResourceScraper` with exact one (implicit) resource.

Follow-up question: why don't you need to create a separate resource per CPU in `cpuscraper` or per disk in `diskscraper`? What makes the processes unique?

Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper.go:65]: 1. For processes, the cardinality can be very high (1000s) which may cause issues when storing the metric data. If grouped by resource, the data is sharded against diff root resources (iiuc).
2. We may want to create lots of metrics against the same process (at the moment its just three, but this will presumably be extended), and its easier to group metrics against the same resource than looking up similar label values.

While they wouldn't be impacted by those issues to the same degree, having a resource per cpu, disk, or network interface could make sense.

@bogdandrutu can probably explain the reasoning in more detail.

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go [hunk 20]
```diff
@@ -32,8 +32,12 @@ import (
 
 // receiver is the type that scrapes various host metrics.
 type receiver struct {
-	config   *Config
-	scrapers []internal.Scraper
+	config *Config
+
+	scrapers         []internal.Scraper
+	resourceScrapers []internal.ResourceScraper
+	allScrapers      []internal.BaseScraper
```
[the remark this answers is not in the document]
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/hostmetrics_receiver.go:39]: Yea good point - updated this to just compute the `allScapers` slice on demand for the `Intialize` & `Close` functions

File: receiver/hostmetricsreceiver/internal/scraper.go [hunk 21]
```diff
@@ -22,15 +22,26 @@ import (
...
+}
+
+// BaseFactory can create a Scraper.
+type BaseFactory interface {
+	// CreateDefaultConfig creates the default configuration for the Scraper.
+	CreateDefaultConfig() Config
+}
+
+// Scraper gathers metrics from the host machine.
+type Scraper interface {
```
[the remark this answers is not in the document]
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper.go:42]: Yea I agree this name is definitely a bit confusing. I'm not sure what to call it though, as I'd prefer to leave "Host" out of the name since it's not an explicit requirement that metrics are scraped from the host - would something like `MetricsSliceScraper` & `ResourceMetricsSliceScraper` be better?

I'll create a separate PR to do this refactor as it will impact all the other scrapers.

File: receiver/hostmetricsreceiver/internal/scraper.go [hunk 22]
```diff
@@ -22,15 +22,26 @@ import (
...
+type BaseScraper interface {
 	// Initialize performs any timely initialization tasks such as
 	// setting up performance counters for initial collection.
 	Initialize(ctx context.Context) error
 	// Close should clean up any unmanaged resources such as
 	// performance counter handles.
 	Close(ctx context.Context) error
+}
+
+// BaseFactory can create a Scraper.
```
[the remark this answers is not in the document]
Comment (james-bebbington) [reply receiver/hostmetricsreceiver/internal/scraper.go]: Changed the comment slightly, though I think the comment below regarding naming the `Scraper` struct will be most helpful to make this less confusing.

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go [hunk 23]
```diff
@@ -119,11 +162,32 @@ func (hmr *receiver) scrapeMetrics(ctx context.Context) {
...
+	}
+
+	if err := hmr.consumer.ConsumeMetrics(ctx, pdatautil.MetricsFromInternalMetrics(metricData)); err != nil {
+		span.SetStatus(trace.Status{Code: trace.StatusCodeDataLoss, Message: fmt.Sprintf("Unable to process metrics: %v", err)})
+		return
+	}
+}
+
+func (hmr *receiver) scrapeAndAppendHostMetrics(ctx context.Context, metricData data.MetricData) error {
 	metrics := internal.InitializeMetricSlice(metricData)
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/hostmetrics_receiver.go:187]:
```go
if len(hmr.hostMetricScrapers) == 0 {
  return nil
}
```
to avoid allocating / sending empty slice

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go [hunk 24]
```diff
@@ -133,21 +197,35 @@ func (hmr *receiver) scrapeMetrics(ctx context.Context) {
...
+		return componenterror.CombineErrors(errors)
 	}
 
-	if metrics.Len() > 0 {
-		err := hmr.consumer.ConsumeMetrics(ctx, pdatautil.MetricsFromInternalMetrics(metricData))
+	return nil
+}
+
+func (hmr *receiver) scrapeAndAppendResourceMetrics(ctx context.Context, metricData data.MetricData) error {
+	rm := metricData.ResourceMetrics()
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/hostmetrics_receiver.go:207]:
```go
if len(hmr.resourceMetricScrapers) == 0) {
  return nil
}
```

File: receiver/hostmetricsreceiver/hostmetrics_receiver.go [hunk 25]
```diff
@@ -160,3 +238,14 @@ func (hmr *receiver) closeScrapers(ctx context.Context) error {
...
 	return nil
 }
+
+func (hmr *receiver) allScrapers() []internal.BaseScraper {
+	allScrapers := make([]internal.BaseScraper, len(hmr.hostMetricScrapers)+len(hmr.resourceMetricScrapers))
+	for i, hostMetricScraper := range hmr.hostMetricScrapers {
+		allScrapers[i] = hostMetricScraper
+	}
+	for i, resourceMetricScraper := range hmr.resourceMetricScrapers {
+		allScrapers[len(hmr.hostMetricScrapers)+i] = resourceMetricScraper
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/hostmetrics_receiver.go:248]: nit: extract `len(hmr.hostMetricScrapers)` to a variable outside of the loop and name it something like `offset`

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 26]
```diff
@@ -0,0 +1,259 @@
...
+// newProcessScraper creates a Process Scraper
+func newProcessScraper(cfg *Config) (*scraper, error) {
+	scraper := &scraper{config: cfg, getProcessHandles: getProcessHandlesInternal}
+
+	var err error
+
+	if len(cfg.Include.Names) > 0 {
+		scraper.includeFS, err = filterset.CreateFilterSet(cfg.Include.Names, &cfg.Include.Config)
+		if err != nil {
+			return nil, fmt.Errorf("error creating process include filters: %v", err)
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:51]:
```go
errors.Wrap(error, "error creating process include filters")
```

File: receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go [hunk 27]
```diff
@@ -0,0 +1,259 @@
...
+
+	var errors []error
+	metadata := make([]*processMetadata, 0, handles.Len())
+	for i := 0; i < handles.Len(); i++ {
+		pid := handles.Pid(i)
+		handle := handles.At(i)
+
+		executable, err := getProcessExecutable(handle)
+		if err != nil {
+			errors = append(errors, fmt.Errorf("error reading process name for pid %v: %v", pid, err))
```
Comment (nilebox) [inline receiver/hostmetricsreceiver/internal/scraper/processscraper/process_scraper.go:141]: `errors.Wrapf(...)`

File: receiver/hostmetricsreceiver/internal/scraper.go (hunk 19 above)
Comment (nilebox) [reply receiver/hostmetricsreceiver/internal/scraper.go:65]: Thanks, that makes perfect sense.
